src/lib/pure_pursuit/PurePursuit.cpp

```diff
@@ -53,43 +54,85 @@ void PurePursuit::updateParams()
+	    || !prev_wp_ned.isAllFinite()) {
 		return NAN;
 	}
 
-	_lookahead_distance = math::constrain(_params.lookahead_gain * vehicle_speed,
-					      _params.lookahead_min, _params.lookahead_max);
-
 	// Pure pursuit
-	const Vector2f curr_pos_to_curr_wp = curr_wp_ned - curr_pos_ned;
+	_lookahead_distance = math::constrain(_params.lookahead_gain * fabsf(vehicle_speed),
+					      _params.lookahead_min, _params.lookahead_max);
+	_curr_pos_to_curr_wp = curr_wp_ned - curr_pos_ned;
 	const Vector2f prev_wp_to_curr_wp = curr_wp_ned - prev_wp_ned;
+	const Vector2f prev_wp_to_curr_pos = curr_pos_ned - prev_wp_ned;
+	const Vector2f prev_wp_to_curr_wp_u = prev_wp_to_curr_wp.unit_or_zero();
+	_distance_along_path = (prev_wp_to_curr_pos * prev_wp_to_curr_wp_u) *
+			       prev_wp_to_curr_wp_u; // Projection of prev_wp_to_curr_pos onto prev_wp_to_curr_wp
+	const Vector2f curr_pos_to_path = _distance_along_path -
+					  prev_wp_to_curr_pos; // Shortest vector from the current position to the path
+	_crosstrack_error = sign(prev_wp_to_curr_wp(1) * curr_pos_to_path(
+					 0) - prev_wp_to_curr_wp(0) * curr_pos_to_path(1)) * curr_pos_to_path.norm();
 
-	if (curr_pos_to_curr_wp.norm() < _lookahead_distance
+	if (_curr_pos_to_curr_wp.norm() < _lookahead_distance
 	    || prev_wp_to_curr_wp.norm() <
 	    FLT_EPSILON) { // Target current waypoint if closer to it than lookahead or waypoints overlap
-		return atan2f(curr_pos_to_curr_wp(1), curr_pos_to_curr_wp(0));
-	}
+		return matrix::wrap_pi(atan2f(_curr_pos_to_curr_wp(1), _curr_pos_to_curr_wp(0)));
 
-	const Vector2f prev_wp_to_curr_pos = curr_pos_ned - prev_wp_ned;
-	const Vector2f prev_wp_to_curr_wp_u = prev_wp_to_curr_wp.unit_or_zero();
-	_distance_on_line_segment = (prev_wp_to_curr_pos * prev_wp_to_curr_wp_u) * prev_wp_to_curr_wp_u;
-	_curr_pos_to_path = _distance_on_line_segment - prev_wp_to_curr_pos;
+	} else {
+
+		if (fabsf(_crosstrack_error) > _lookahead_distance) { // Path segment is outside of lookahead (No intersection point)
+			const Vector2f prev_wp_to_closest_point_on_path = curr_pos_to_path + prev_wp_to_curr_pos;
+			const Vector2f curr_wp_to_closest_point_on_path = curr_pos_to_path - _curr_pos_to_curr_wp;
+
+			if (prev_wp_to_closest_point_on_path * prev_wp_to_curr_wp <
+			    FLT_EPSILON) { // Target previous waypoint if closest point is on the the extended path segment "behind" previous waypoint
+				return matrix::wrap_pi(atan2f(-prev_wp_to_curr_pos(1), -prev_wp_to_curr_pos(0)));
+
+			} else if (curr_wp_to_closest_point_on_path * prev_wp_to_curr_wp >
+				   FLT_EPSILON) { // Target current waypoint if closest point is on the extended path segment "ahead" of current waypoint
+				return matrix::wrap_pi(atan2f(_curr_pos_to_curr_wp(1), _curr_pos_to_curr_wp(0)));
+
+			} else { // Target closest point on path
+				return matrix::wrap_pi(atan2f(curr_pos_to_path(1), curr_pos_to_path(0)));
+			}
+
+
+		} else {
```

Comment:
Same, does not need else as the previous if-elseif-else is catch all with returns